source/server/connection_handler_impl.cc

```diff
@@ -54,6 +66,30 @@ void ConnectionHandlerImpl::removeListeners(uint64_t listener_tag) {
   }
 }
 
+void ConnectionHandlerImpl::removeFilterChains(
+    const Network::DrainingFilterChains& draining_filter_chains, std::function<void()> completion) {
+  for (auto& listener : listeners_) {
+    // TODO(lambdai): merge the optimistic path and the pessimistic locking.
```

Comment:
I'm not sure what this means. Can you expand. Also, can you make it clear why you are using a deferred task to run the completion? It seems like the fact that the completion can't run in the same callstack is a caller implementation detail and should be handled there potentially?